### PR TITLE
Increase efficiency for large playlist

### PR DESCRIPTION
**Pull request by hori-ryota**
Increase efficiency for large playlist.

before

```
BenchmarkDecodeMediaPlaylist-8        50          23445112 ns/op        12282139 B/op     160240 allocs/op
BenchmarkEncodeMediaPlaylist-8       100          14378131 ns/op         1749294 B/op      81643 allocs/op
```

after

```
BenchmarkDecodeMediaPlaylist-8       100          16781361 ns/op        11001559 B/op     120172 allocs/op
BenchmarkEncodeMediaPlaylist-8       100          13064495 ns/op         1744527 B/op      81247 allocs/op
```

### CODE REVIEW

**Comment by hori-ryota** [comment]
I want to use this. Would you merge this?

**Comment by bradleyfalzon** [comment]
This looks good at first glance, will need to spend some time with it, under a little time pressure with others things, maybe a week if not sooner.

It'll help me if you also describe what you did and why.

Thanks again, looking forward to reviewing it.

**Comment by bradleyfalzon** [comment]
Unless @grafov has time or further thoughts.

**Comment by hori-ryota** [comment]
I'm sorry, I have hurried too much.

Add more code, before
```
BenchmarkDecodeMasterPlaylist-8    30000             44103 ns/op           14958 B/op         77 allocs/op
BenchmarkDecodeMediaPlaylist-8        50          23445112 ns/op        12282139 B/op     160240 allocs/op
BenchmarkEncodeMediaPlaylist-8       100          14378131 ns/op         1749294 B/op      81643 allocs/op
```
after
```
BenchmarkDecodeMasterPlaylist-8    50000             26881 ns/op           12232 B/op         38 allocs/op
BenchmarkDecodeMediaPlaylist-8       100          16190399 ns/op         9200543 B/op      80021 allocs/op
BenchmarkEncodeMediaPlaylist-8       300           3878858 ns/op           35212 B/op        269 allocs/op
```

I'm using forked repository now, if you like this please give me an opinion.

**Comment by bradleyfalzon** [comment]
OK, this will take me longer to analyse. A few quick comments before 

There appears to be many improvements here, which is excellent, but some commits are just fixes to bugs caused by other commits, and also some breaking changes we almost certainly cannot accept (unexporting `MediaSegment.Duration` - and I'm not sure why).

The benchmark itself seems to require an extra 1MB, is it to run the benchmark a hundred times on a smaller example to get the same benchmark result?

It's likely we won't accept the PR without at least some rebasing, instead opting to break the changes down into more manageable chunks.

Can you describe generally the various changes you've made, for example:

- Add decoding benchmark
- Move common case statements to top for quicker decoding
- Switch to strconv.Parse* functions instead of fmt.Sscanf
- Preallocate some buffers 

Thank you, some of these changes are excellent and I look forward to merging them.

File: reader.go [hunk 1]
```diff
@@ -74,27 +57,41 @@ func (p *MasterPlaylist) decode(buf *bytes.Buffer, strict bool) error {
 	if strict && !state.m3u {
 		return errors.New("#EXT3MU absent")
 	}
+
+	p.buf = *bytes.NewBuffer(make([]byte, 0, buf.Len()*2))
```
Comment (bradleyfalzon) [inline reader.go]: What's this change for? I don't think `p.buf` is required anymore?

**Comment by bradleyfalzon** [comment]
OK,

So here's the benchmarks for each commit where the tests pass:

Before: 652e4af8ba5f7a4c7b14e0b7c560f53e6b7bbeff
```
go test -bench . -benchtime 10s -benchmem
testPASS
BenchmarkDecodeMediaPlaylist-4       500          24428571 ns/op        12279906 B/op     160060 allocs/op
BenchmarkEncodeMediaPlaylist-4      1000          13285791 ns/op         1614948 B/op      80166 allocs/op
ok      github.com/hori-ryota/m3u8      29.289s
```

Very good: f4f36c983c56a5d503b194044b4082afc3091bc2
```
go test -bench . -benchtime 10s -benchmem
testPASS
BenchmarkDecodeMediaPlaylist-4      1000          17009632 ns/op        10999885 B/op     120060 allocs/op
BenchmarkEncodeMediaPlaylist-4      1000          12839038 ns/op         1614470 B/op      80126 allocs/op
ok      github.com/hori-ryota/m3u8      32.921s
```

No substantial benefit: 52afbdf730146a76302aae3defdecbf846638836
```
go test -bench . -benchtime 10s -benchmem
testPASS
BenchmarkDecodeMediaPlaylist-4      1000          17296504 ns/op         8876106 B/op     120052 allocs/op
BenchmarkEncodeMediaPlaylist-4      1000          12718539 ns/op         1608931 B/op      80126 allocs/op
ok      github.com/hori-ryota/m3u8      33.055s
```

No substantial benefit: 78531af5c52f8949357ad466fe9a87bc51570f66
```
go test -bench . -benchtime 10s -benchmem
testPASS
BenchmarkDecodeMediaPlaylist-4      1000          17009632 ns/op        10999885 B/op     120060 allocs/op
BenchmarkEncodeMediaPlaylist-4      1000          12839038 ns/op         1614470 B/op      80126 allocs/op
ok      github.com/hori-ryota/m3u8      32.921s
```

Very good: efb283588847223639447511790331eee9c3ba56
```
go test -bench . -benchtime 10s -benchmem
testPASS
BenchmarkDecodeMediaPlaylist-4      1000          15421930 ns/op         8559427 B/op      80033 allocs/op
BenchmarkEncodeMediaPlaylist-4      1000          12535349 ns/op         1609807 B/op      80085 allocs/op
```

Additional benchmark: e89f4652aaa1d81e0eed4c809cecb94f97a14cac
```
go test -bench . -benchtime 10s -benchmem
testPASS
BenchmarkDecodeMasterPlaylist-4  1000000             23426 ns/op           14955 B/op         77 allocs/op
BenchmarkDecodeMediaPlaylist-4      1000          15394268 ns/op         8559422 B/op      80033 allocs/op
BenchmarkEncodeMediaPlaylist-4      1000          12786649 ns/op         1609807 B/op      80085 allocs/op
ok      github.com/hori-ryota/m3u8      54.993s
```

Very good `EncodeMediaPlaylist`, but regression in `DecodeMediaPlaylist` and break unexports a field? Multiple commits here and all break tests so haven't been individually benchmarked.

Results for changes between e89f4652aaa1d81e0eed4c809cecb94f97a14cac and 3629b3ce025e6cef5867017ae121e9e9dbbe5d47
```
go test -bench . -benchtime 10s -benchmem
testPASS
BenchmarkDecodeMasterPlaylist-4  1000000             22135 ns/op           14712 B/op         73 allocs/op
BenchmarkDecodeMediaPlaylist-4      1000          16163201 ns/op         9200526 B/op      80019 allocs/op
BenchmarkEncodeMediaPlaylist-4      5000           3692252 ns/op            2116 B/op         19 allocs/op
ok      github.com/hori-ryota/m3u8      59.112s
```

Very good: 5f14343349bd0f71b1745f8452fc4ea3d97fe641
```
go test -bench . -benchtime 10s -benchmem
testPASS
BenchmarkDecodeMasterPlaylist-4  1000000             10280 ns/op           12232 B/op         38 allocs/op
BenchmarkDecodeMediaPlaylist-4      1000          16184904 ns/op         9200524 B/op      80019 allocs/op
BenchmarkEncodeMediaPlaylist-4      5000           3674522 ns/op            2116 B/op         19 allocs/op
ok      github.com/hori-ryota/m3u8      47.063s
```

No substantial improvement: 8b5b7d07f07c004fdde31f18e03f71ea372208c8

```
go test -bench . -benchtime 10s -benchmem
testPASS
BenchmarkDecodeMasterPlaylist-4  1000000             10191 ns/op           12232 B/op         38 allocs/op
BenchmarkDecodeMediaPlaylist-4      1000          16525064 ns/op         9200524 B/op      80019 allocs/op
BenchmarkEncodeMediaPlaylist-4      5000           3682421 ns/op            2116 B/op         19 allocs/op
ok      github.com/hori-ryota/m3u8      47.339s
```

No substantial improvement: fe8dbb235c735869d83c3c6058c74a5c0f624178

```
go test -bench . -benchtime 10s -benchmem
testPASS
BenchmarkDecodeMasterPlaylist-4  2000000             10417 ns/op           12240 B/op         35 allocs/op
BenchmarkDecodeMediaPlaylist-4      1000          16498553 ns/op         9200524 B/op      80019 allocs/op
BenchmarkEncodeMediaPlaylist-4      5000           3666946 ns/op            2116 B/op         19 allocs/op
ok      github.com/hori-ryota/m3u8      67.854s
```

**Comment by bradleyfalzon** [comment]
There are some changes here that are very good, I don't want to merge them all as some performance improvements need to be weighed up against the increase risk of bugs and the readability.

I'll continue to look, but there's a lot to go through here.

**Comment by bradleyfalzon** [comment]
OK, I've had a chance to look through all the changes and again, would like to thank you for the time and effort you've put into these proposals. This is much appreciated and helps others, so thank you.

Many of these changes I'd like to incorporate, but the PR as it is:
- Has too many commits, some changes should be rebased
- Some broken tests and then fixes should be rebased into the one commit, such as 43afc6d37625e6dd0728b4d74456f8d5efb47a75, 3629b3ce025e6cef5867017ae121e9e9dbbe5d47, 5f14343349bd0f71b1745f8452fc4ea3d97fe641, 8b5b7d07f07c004fdde31f18e03f71ea372208c8
- Some improvements don't provide a large benefit for the amount of code they're changing, such as 52afbdf730146a76302aae3defdecbf846638836 and 52afbdf730146a76302aae3defdecbf846638836
- The changes to the Duration field break backwards compatibility, and could potentially be solved via other methods, see a98d61c9643bd2c0f2a2ff4fe2cdb2d37f492e4d

But the general improvements are good, I've proposed another PR #58 with benchmarks comparing your PR with an alternative as well as including benchmarks before and after each test.

It doesn't quite get to the same level yours did, but it's substantially less changes, doesn't break backwards compatibility and still comes close. It doesn't incorporate improvements in 5f14343349bd0f71b1745f8452fc4ea3d97fe641 - as this added a lot of complication and wasn't clear to me where the benefits in the real world would be.

The alternative PR #58 is simply a discussion point, if you'd like me to review some of your improvements, point them out exactly and I'll let you know whether we could incorporate them.

@grafov I'd love some of your input in this as well, as it's adding some complication, I tried to minimise it as much as I could.

**Comment by hori-ryota** [comment]
Thanks, your PR is nice! I will use that as reference from now on.

**Comment by grafov** [comment]
@hori-ryota thank for starting this topic about optimization! Nice results with your patches.
@bradleyfalzon I agree with you way, let optimize code step by step. So let merge #58. 
I think we can include later changes proposed in 5f14343. As I remember logic with a regexp was added to the code without benchmarking just for making this part of code shorter. It was a disputable solution.

**Comment by bradleyfalzon** [comment]
Thanks both, I've reverted a small change that had no improvement after all (see 78531af5c52f8949357ad466fe9a87bc51570f66), and ever so slightly rewritten the duration cache.

**Comment by bradleyfalzon** [comment]
OK, this has been merged.

@hori-ryota can you checkout master and determine whether you're satisfied with the performance improvements, or if there's other changes you had proposed that we reinvestigate?

Note to self: I'd still like to look at the following at a later date (I think we maybe able to do the allocation in NewMediaPlaylist too), but it may also just be a good idea to document how to do this via AppendSegment methods:
```
@@ -288,10 +288,18 @@ func (p *MediaPlaylist) Remove() (err error) {
  	return nil
  }
  
 +var segmentStore []MediaSegment
 +var segmentStoreTail = 0
 +
  // Append general chunk to the tail of chunk slice for a media playlist.
  // This operation does reset playlist cache.
  func (p *MediaPlaylist) Append(uri string, duration float64, title string) error {
 -	seg := new(MediaSegment)
 +	if segmentStoreTail == len(segmentStore) {
 +		segmentStore = make([]MediaSegment, p.capacity)
 +		segmentStoreTail = 0
 +	}
 +	seg := &segmentStore[segmentStoreTail]
 +	segmentStoreTail++
  	seg.URI = uri
  	seg.Duration = duration
  	seg.Title = title
```

**Pull request closed without merging**